### PR TITLE
Timelock Deploy Script

### PR DESCRIPTION
**Pull request by ElliotFriedman**

### CODE REVIEW

**Review by k-xo — COMMENTED** [review]
This is looking good, had a suggestion of maybe something we could do rather than replacing  addresses in the env variable for each network as we might be likelier to run into a mistake like that.

```
  const network = hre.network.name;

  const proposerAddresses = {
    arbitrum: [],
    mainnet: []
  };

await (await ethers.getContractFactory('')).deploy(core, oneDay, proposerAddresses[network], executors);
```
perhaps we could do something like this, and similarly so for executor addresses

File: scripts/timelockDeploy.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  await timelock.deployed();
+
+  console.log(`\nTimelock deployed to: ${timelock.address}`);
+
+  const TIMELOCK_ADMIN_ROLE = await timelock.TIMELOCK_ADMIN_ROLE();
+  let tx = await timelock.grantRole(TIMELOCK_ADMIN_ROLE, multisig);
+  await tx.wait(1);
+
+  tx = await timelock.renounceRole(TIMELOCK_ADMIN_ROLE, deployer.address);
+  await tx.wait(1);
```
Comment (Namaskar-1F64F) [inline scripts/timelockDeploy.ts:31]: Not positive on the syntax here, but here's another way to to waiting without making a temp variable.
```suggestion
  await (
    await timelock.renounceRole(TIMELOCK_ADMIN_ROLE, deployer.address)
  ).wait();
```

**Review by Namaskar-1F64F — APPROVED** [review]
looks good to me

**Review by RusseII — APPROVED** [review]
Woo!

File: scripts/timelockDeploy.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+/// deploying the timelock on arbitrum should look like this:
+///   MSIG=0x1A1075cef632624153176CCf19Ae0175953CF010 PROPOSERS=0x1A1075cef632624153176CCf19Ae0175953CF010,0xA96D4a5c343d6fE141751399Fc230E9E8Ecb6fb6,0xB320e376Be6459421695F2b6B1E716AE4bc8129A,0xd90E9181B20D8D1B5034d9f5737804Da182039F6 EXECUTORS=0x1A1075cef632624153176CCf19Ae0175953CF010 \
+///   npx hardhat run --network arbitrumOne scripts/timelockDeploy.ts
+
+/// Script does not hardcode msig, proposers or executors because this system will be deployed
+/// onto multiple networks where all of these addresses will change
+
+async function deploy(proposers: string[], executors: string[], multisig: string) {
+  const deployer = (await ethers.getSigners())[0];
+  const timelock = await (await ethers.getContractFactory('TimelockController')).deploy(oneDay, proposers, executors);
```
[the remark this answers is not in the document]
Comment (ElliotFriedman) [reply scripts/timelockDeploy.ts:21]: yes

File: protocol-configuration/mainnetAddresses.ts [hunk 3]
```diff
@@ -192,7 +210,13 @@ const MainnetAddresses: MainnetAddresses = {
   arbitrumOptimisticTimelock: {
     address: '0x980A05105a53eCa7745DA40DF1AdE6674fc73eD5',
     artifactName: 'OptimisticTimelock',
-    category: AddressCategory.Core,
+    category: AddressCategory.Deprecated,
```
[the remark this answers is not in the document]
Comment (ElliotFriedman) [reply protocol-configuration/mainnetAddresses.ts:213]: where do we check that a contract gets used? would definitely make sense to log a warning if you reference a deprecated smart contract

File: scripts/timelockDeploy.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  const { EXECUTORS, PROPOSERS, MSIG } = process.env;
+
+  if (!EXECUTORS || !PROPOSERS || !MSIG) {
+    throw new Error('Proposer(s), Executor(s), or Multisig not set in environment variables');
+  }
+
+  /// if there is an array, break addresses into their respective elements
+  const proposers = PROPOSERS.split(',');
+  const executors = EXECUTORS.split(',');
+
```
[the remark this answers is not in the document]
Comment (ElliotFriedman) [reply scripts/timelockDeploy.ts:65]: good point, by going with @k-xo's solution, I was able to remove the use of env variables

**Review by RusseII — APPROVED** [review]

**Review by k-xo — APPROVED** [review]

**Review by k-xo — APPROVED** [review]
LGTM

**Pull request merged**